crates/turborepo-lib/src/shim.rs

```diff
@@ -176,50 +177,237 @@ struct PackageJson {
+    fn generate_nested_path(root_path: &Path) -> Option<PathBuf> {
+        Some(
+            root_path
+                .join("node_modules")
+                .join("turbo")
+                .join("node_modules"),
+        )
+    }
 
-        let local_turbo_package_path = repo_root
-            .join("node_modules")
-            .join("turbo")
+    // Linked strategy:
+    // - `pnpm install`
+    // - `npm install --install-strategy=linked`
+    fn generate_linked_path(root_path: &Path) -> Option<PathBuf> {
+        fs_canonicalize(root_path.join("node_modules").join("turbo").join("..")).ok()
+    }
+
+    // The unplugged directory doesn't have a fixed path.
+    fn get_unplugged_base_path(root_path: &Path) -> PathBuf {
+        let yarn_rc_filename =
+            env::var_os("YARN_RC_FILENAME").unwrap_or_else(|| OsString::from(".yarnrc.yml"));
+        let yarn_rc_filepath = root_path.join(yarn_rc_filename);
+
+        let yarn_rc_yaml_string = fs::read_to_string(yarn_rc_filepath).unwrap_or_default();
+        let yarn_rc: YarnRc = serde_yaml::from_str(&yarn_rc_yaml_string).unwrap_or_default();
+
+        root_path.join(yarn_rc.pnp_unplugged_folder)
+    }
+
+    // Unplugged strategy:
+    // - berry 2.1+
+    fn generate_unplugged_path(root_path: &Path) -> Option<PathBuf> {
+        let platform_package_name = TurboState::platform_package_name();
+        let unplugged_base_path = Self::get_unplugged_base_path(root_path);
+
+        unplugged_base_path
+            .read_dir()
+            .ok()
+            .and_then(|mut read_dir| {
+                // berry includes additional metadata in the filename.
+                // We actually have to find the platform package.
+                read_dir.find_map(|item| match item {
+                    Ok(entry) => {
+                        let file_name = entry.file_name();
+                        if file_name
+                            .to_string_lossy()
+                            .starts_with(&platform_package_name)
+                        {
+                            Some(unplugged_base_path.join(file_name).join("node_modules"))
+                        } else {
+                            None
+                        }
+                    }
+                    Err(_) => None,
+                })
+            })
+    }
+
+    // We support six per-platform packages and one `turbo` package which handles
+    // indirection. We identify the per-platform package and execute the appropriate
+    // binary directly. We can choose to operate this aggressively because the
+    // _worst_ outcome is that we run global `turbo`.
+    //
+    // In spite of that, the only known unsupported local invocation is Yarn/Berry <
+    // 2.1 PnP
+    pub fn infer(root_path: &Path) -> Option<Self> {
+        let platform_package_name = TurboState::platform_package_name();
+        let binary_name = TurboState::binary_name();
+
+        let platform_package_json_path = Path::new("")
```

Comment:
It might be better to convert from an array according to the [path docs](https://doc.rust-lang.org/std/path/index.html)

Reply:
This is another case where we really just need a macro. Absolutely no reason to do any of this at runtime. The closest crate is this: https://github.com/zertosh/path_macro, which is a dtolnay special, but extracted.

But that one isn't actually good enough for me.

Reply:
(Updated.)